clang/test/CodeGenSYCL/esimd-accessor-ptr-md.cpp

```diff
@@ -0,0 +1,30 @@
+// RUN: %clang_cc1 -fsycl -fsycl-explicit-simd -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -emit-llvm %s -o - | FileCheck %s
+
+// This test checks proper metadata generation for accessors used in ESIMD
+// kernels:
+// 1) Proper 'kernel_arg_accessor_ptr' metadata is generated by the FE for ESIMD
+//    kernels
+// 2) Pointers originating from accessors are marked with 'buffer_t' and proper
+//    argument kind.
+
+#include "Inputs/sycl.hpp"
+
+void test(int val) {
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read_write> accessorA;
+  cl::sycl::accessor<int, 1, cl::sycl::access::mode::read> accessorB;
+
+  cl::sycl::kernel_single_task<class esimd_kernel>(
+      [=]() __attribute__((sycl_explicit_simd)) {
+        accessorA.use(val);
+        accessorB.use();
+      });
+  // --- Name
+  // CHECK: define spir_kernel void @_ZTSZ4testiE12esimd_kernel(
+  // --- Signature
+  // CHECK:   i32 addrspace(1)* nocapture "VCArgumentDesc"="buffer_t" "VCArgumentIOKind"="0" "VCArgumentKind"="2" %_arg_,
+  // CHECK:   i32 "VCArgumentDesc" "VCArgumentIOKind"="0" "VCArgumentKind"="0" %_arg_1,
+  // CHECK:   i32 addrspace(1)* nocapture "VCArgumentDesc"="buffer_t" "VCArgumentIOKind"="0" "VCArgumentKind"="2" %_arg_3)
+  // --- Attributes
+  // CHECK: {{.*}} !kernel_arg_accessor_ptr ![[ACC_PTR_ATTR:[0-9]+]] !sycl_explicit_simd !{{[0-9]+}} {{.*}}{
```

Comment:
Could you please check that `__init_esimd` is called instead of regular `__init` in case of ESIMD kernel?

Reply:
This is basically proven by the fact that an accessor is translated into a single pointer arg (as in __init_esimd) rather than 4 (as in __init). I'll see if I can make direct check. Do you have suggestions, BTW?

Reply:
I don't remember exactly, but I was under impression that we don't enable early optimizations for ESIMD mode, am I right?
In case if IR is not optimized, I would just check that there are **two** calls to __init_esimd functions, i.e. (the following LLVM IR may contain errors):
```
CHECK: call spir_func void @{{.*}}__init_esimd{{.*}}(%"class{{.*}}cl::sycl::accessor" addrspace(4)* {{.*}}, i32 addrspace(1)* {{.*}})
```
Or you can even check which argument of the kernel is passed like here **https://github.com/intel/llvm/blob/3b8af2229eeb45f26b610d5e33a007fc6652532e/clang/test/CodeGenSYCL/accessor_inheritance.cpp#L84** .

In case if IR is optimized for ESIMD mode, we can check presence of `__init_esimd` in non-optimized IR.

Reply:
ok, thanks. Will try